scripts/plugin_list.py

```diff
@@ -0,0 +1,85 @@
+import datetime
+import pathlib
+import re
+
+import packaging.version
+import requests
+import tabulate
+
+FILE_HEAD = r"""Plugins List
+============
+
+The plugins are listed automatically from PyPI.
+Only PyPI projects that match "pytest-\*" are considered plugins.
+Packages classified as inactive are also excluded.
+
+"""
+DEVELOPMENT_STATUS_CLASSIFIERS = (
+    "Development Status :: 1 - Planning",
+    "Development Status :: 2 - Pre-Alpha",
+    "Development Status :: 3 - Alpha",
+    "Development Status :: 4 - Beta",
+    "Development Status :: 5 - Production/Stable",
+    "Development Status :: 6 - Mature",
+    "Development Status :: 7 - Inactive",
+)
+
+
+def iter_plugins():
+    regex = r">([\d\w-]*)</a>"
+    response = requests.get("https://pypi.org/simple")
+    for match in re.finditer(regex, response.text):
+        name = match.groups()[0]
+        if not name.startswith("pytest-"):
+            continue
+        response = requests.get(f"https://pypi.org/pypi/{name}/json")
+        if not response.ok:
+            continue
+        info = response.json()["info"]
+        if "Development Status :: 7 - Inactive" in info["classifiers"]:
+            continue
+        for classifier in DEVELOPMENT_STATUS_CLASSIFIERS:
+            if classifier in info["classifiers"]:
+                status = classifier[22:]
+                break
+        else:
+            status = "NA"
+        requires = "NA"
+        if info["requires_dist"]:
+            for requirement in info["requires_dist"]:
+                if requirement == "pytest" or "pytest " in requirement:
+                    requires = requirement
+                    break
+        releases = response.json()["releases"]
+        for release in sorted(releases, key=packaging.version.parse, reverse=True,):
+            if releases[release]:
+                release_date = datetime.date.fromisoformat(
+                    releases[release][-1]["upload_time_iso_8601"].split("T")[0]
+                )
+                last_release = release_date.strftime("%b %d, %Y")
+        name = f'`{info["name"]} <{info["project_url"]}>`_'
+        pyversions = f'.. image:: https://img.shields.io/pypi/pyversions/{info["name"]}'
```

Comment:
The images have varying sizes on the page:

![image](https://user-images.githubusercontent.com/1085180/96750752-b6f3e900-13a2-11eb-8593-ac94f61c708c.png)


Any idea on how to have them all their original sizes instead?

Reply:
Currently the shields are vector graphics be default. Maybe that can be fixed.

Reply:
1.1 One solution could be to scrape the dimensions of the shield from the respective link and set the dimensions explicitly using the width and height options of the image directive. This would add 2 lines to the file for each image. Using inline-notation so that the rst is more human readable would mean 3 more lines essentially making the file 4 times as large.
This would also double the requests to websites made by the script.
1.2 Instead of making the request one could manually parse the classifiers and based on the resulting list the dimensions could be defined as well. Parsing of the metadata is part of what makes shields nice though.

2. Shields.io also offers the badges in the PNG format, but the images are not as sharp of course. See for example  the [pytest-django pyversions shield (PNG)](https://raster.shields.io/pypi/pyversions/pytest-django).

Which option preferred? Any other ideas?

Reply:
Hmm as @bluetech mentions in https://github.com/pytest-dev/pytest/pull/7921#pullrequestreview-515705510, actually just referencing the outside images is problematic in this case. Perhaps the script can download the images when running and store them into the pytest's repo?

Reply:
Do you mean having something like a `pyversions` folder in the docs with one shield for each of the almost 800 plugins or something closer to `1.2` of my previous comment? Storing the shields in the repo is probably a must either way.

If the same image is embedded multiple times in a single html document, will there be multiple requests for the same image?

Reply:
> One solution could be to scrape the dimensions of the shield from the respective link and set the dimensions explicitly using the width and height options of the image directive. This would add 2 lines to the file for each image. Using inline-notation so that the rst is more human readable would mean 3 more lines essentially making the file 4 times as large.

If it looks nice, I don't think the length of the rst file is an issue, users probably won't be reading it anyway, only looking at the rendered page.

> Do you mean having something like a pyversions folder in the docs with one shield for each of the almost 800 plugins or something closer to 1.2 of my previous comment?

I meant the former. Shouldn't be an issue as they are all very small SVG files.

> If the same image is embedded multiple times in a single html document, will there be multiple requests for the same image?

If they share the same filename, the browser makes only a single request.

Reply:
I will have a go at implementing it so that the badges are part of the pytest docs and referenced locally. Quite busy atm, but should be done within a week.

Reply:
800 checked in and rapidly changing ~essentially binary files does not seem like a good idea to me.  it will quickly become a significant portion of the git history size

Reply:
An update script could deduplicate them before adding them to the repo - but I wouldn't find it too crazy to let git take care of that either.

Out of curiosity, I downloaded all 780 of them. They have a total size of 3.1 MB (1.1 to 1.2 KB each), but there are only 77 deduplicated variants.

If I commit them all without manual deduplication, I end up with a `.git` which is 240K - compared with 164K for an empty git repo (!) or 25M for pytest's current `.git`, that really doesn't seem like a lot. Also I don't see how they'd be rapidly changing - I'd expect a change once all couple of months.

Reply:
they contain the package version (probably changes every few weeks) and the target python versions (probably changes twice a year) -- I think that's pretty often

Reply:
Can't we just capture their text, and generate them on the spot in the docs build process with something like `anybadge`?  
Or we can just have a column for each python version supported by the pytest and just put `X` or `✓` there

Reply:
> they contain the package version 

I don't see a package version:

![image](https://user-images.githubusercontent.com/625793/97545310-1a959c00-19cb-11eb-94ef-88669d2ac864.png)

Reply:
> Or we can just have a column for each python version supported by the pytest and just put X or ✓ there

This can actually be a solution to another issue, which is the varying size of the columns because to the image dimensions.

Reply:
Storing the images locally was easy to implement, but i can't come up with a solution for the image size.
There are packages with tons of "compatible" python versions. Check out [pytest-pythonpath](https://img.shields.io/pypi/pyversions/pytest-pythonpath) and [pytest-dotenv](https://img.shields.io/pypi/pyversions/pytest-dotenv) and have a look the the rendered docs for the PR. The badges just get crazy long and are impossible to fit onto the page, not even talking about the column. Setting the dimensions of the images in the rst did not really work, but i fixed the scaling issue by setting the columnwidth to the biggest image width. This however means that half the table is the column for the badges, because some are so incredibly white. This means, that it is mostly whitespace, just to accomodate the outliers. Here are some ideas for how to handle this:
1. Do not show compatible python versions at all.
2. Limit the "relevant" versions only show relevant versions in the badge. This is easy, since the badges are already stored locally. I parsed the title of the shield, extracted the versions and then used the api to download unique version combinations into badges named something like `35_36_37_38_39.svg` into `doc/en/img/pyversions/`. We could just ignore anything below 3.5 (current minimal compatble python for pytest) for example. This would arguably still be to long.
3. Just put a link to the shield instead of embedding it.

I am in favor of option 1, since anyone interested can also go to the pypi page of the plugin, which is linked to already and the compatible pytest version can be used for a rough estimate, whether the plugin would be useful to the user.

Reply:
Thinking about it some more, :+1: from me as well for getting rid of the column entirely. Like you say, if a plugin supports up-to-date pytest versions, I guess the answer to "will this support Python version xyz I'm interested in" is rather likely to be "yup!"

Reply:
I removed the column. Reimplementing some kind of pyversion compatibility can be done in a future PR if a manageable way can be found. Resolved!